### ISSUE TITLE
GPS error of over 150m

### ISSUE THREAD

**Issue by ghuotvez** [NRCan/GSC-Field-Application#79]
It seems like this bug is found on all tested devices. All being on W10 build 1909. The whole team have experienced it.

@jameljoseph @eygirard @rcocking @gisgorilla

**Comment by ghuotvez** [comment]
@eygirard says that the GPS is fine on a Trimble 7 tablet. Seems like the problem is coming from Panasonic tablets...

**Comment by ghuotvez** [comment]
After a few discussion with an IT guy from Panasonic @eygirard tested out the GPS with a non UWP application (Arc Pad in this case) and the accuracy is fine.

So the problem lies somewhere else and it comes from Windows. They might have introduced something new and we'll need to check this out. I'll get my hands on a tablet and I'll try to debug this while being connected by bluetooth. There is a how-to description in the installation guide on gcdoc @jameljoseph  if you have the time to check it out before I do.

**Comment by jameljoseph** [comment]
Sure I will check it out.  I checked today and I got 232 m accuracy even though it pin pointed me in the correct spot and my sons school.



Sent from my Bell Samsung device over Canada's largest network.



-------- Original message --------

**Comment by ghuotvez** [comment]
@jameljoseph  That's quite intriguing, could it be the location is fine but the accuracy value we are getting is pooched?

**Comment by ghuotvez** [comment]
Answer from them:

>If these units have both Sierra Wireless Location Sensor and u-blox GNSS Location Sensor, does it make a difference if you switch sensors?  The u-blox GNSS is wired to be able to use external antennas, is slightly more accurate and faster to first-fix.

>If this app is able to communicate directly with the GPS Com port, my suggestion is to use this method instead of the Windows Location Sensor.  This way, the software communicates directly with the GPS hardware.

>Windows Location Sensors also make use of IP address to determine location, which is beyond our control and the functionality may differ slightly depending on which Windows build you’re using.

**Comment by jameljoseph** [comment]
Before my cursor was off but lately it is dead on, even though the cursor is in the right spot.

 I will  try at maxs soccer practice tonight and see if it is dead on as well



Sent from my Bell Samsung device over Canada's largest network.



-------- Original message --------

**Comment by ghuotvez** [comment]
After a quick look with one panasonic it seems like the uwp app can't access the right com port, as opposed to arc pad and panasonic utility gps viewer. Found this gem that might explain our problem...

https://docs.microsoft.com/en-us/answers/questions/8363/uwp-access-to-com3-serial-port-on-panasonic-tablet.html

**Comment by jameljoseph** [comment]
Good job.  I will check it out here shortly



Sent from my Bell Samsung device over Canada's largest network.



-------- Original message --------

**Comment by ghuotvez** [comment]
I was able to get a 1m accuracy! Although that's because I was able to make the tablet read the gps from a bluetooth connected device (xgps160 dual sky pro).

First I paired the device. 

Then W10 doesn't have any drivers for it. I found a way of getting them and bypassing W10 system location service.
I downloaded this (which includes the proper drivers):
https://www.gpssensordrivers.com/

Then I was able to select and install driver for the xgps160 device. This little executable also enables you to sniff at all sensors on the tablet (accelerometer, light, ... and gps) I could see that it always reverts to wifi by default. Went on and activated the localisation with the bluetooth and saw it take the satellite and push a good location. Opened GSC Field app and it was working well. Although GPSComplete always needs to stay opened for the W10 location service to be bypassed, I'll try and force it to use com5 which is actually where the tablet gps chip is set to be readable from.

**Comment by ghuotvez** [comment]
![image](https://user-images.githubusercontent.com/19910413/100637800-47f2a400-3301-11eb-8b7a-ffa2217dbb5a.png)

**Comment by ghuotvez** [comment]
Stumbled upon a Panasonic tablet with W10 version 1607 with a field app dating from this time. GPS works fine ... like in the good ol' days...

**Comment by ghuotvez** [comment]
@eygirard 's got one device in 1909 that works well. The device was updated to that version and simply worked fine...the mystery keeps on.

**Comment by rcocking** [comment]
Great news!

**Comment by ghuotvez** [comment]
> Great news!

Yes indeed! We'll be able to ghost that device and install this version on others to make sure it's fine.

**Comment by jameljoseph** [comment]
I still cannot get it to work on my devices!!

**Comment by ghuotvez** [comment]
> I still cannot get it to work on my devices!!

Yeah pretty much the same here. No idea why Étienne's tablet suddenly started working fine after an update.

**Comment by ghuotvez** [comment]
I got a tablet with the latest updated W10, and a new package that I created. I was also able to remotely connect to the tablet via bluetooth to debug it and it seems like the position change event occurs at initialization only and that's it it never gets triggered again. Of course if it's never updated then first position will not be great.

Before going in debug mode via bluetooth I tested this https://docs.microsoft.com/en-us/uwp/api/windows.devices.serialcommunication?view=winrt-19041 within package manifest, hoping it would get proper access to com5 port on the tablet. Didn't do much like I thought, since we're not using any of the libraries that interact with this capability.

**Comment by ghuotvez** [comment]
There seems to be changed made on how to call for the position: https://docs.microsoft.com/en-us/windows/uwp/maps-and-location/get-location 

MS code doesn't look like what is in master branch.

**Comment by ghuotvez** [comment]
That is really weird. I was able to get the code on debug via bluetooth on a tablet. It refreshes fine, it's only that it always goes to wifi location by default, even though there is a GPS port like u-blox. Then I tested a bluetooth GPS via my laptop, I get the 1m accuracy but it always flip between satellite position and wifi, like if windows was enforcing wifi for some reasons. To make the skypro gps work I had to download gpscomplete.exe 64 and install proper drivers for it then the laptop was able to get a fix.

**Comment by ghuotvez** [comment]
Something that I've found on u-blox support center
https://portal.u-blox.com/s/question/0D52p00008NKMLaCAP/panasonic-toughpad-does-not-update-position-when-moving

I downloaded and tried u-center to sniff at what is going on in the back, and it seems fine. Really it seems like a Windows thing on old devices. From the comments it seems like new devices don't have this problem as opposed to old one that were upgrade to W10. That might be our problem...

**Comment by ghuotvez** [comment]
I've done some changes within the app, I removed the threshold at which we're expecting an update from location, this seems to have fixed the "always" revert to wifi, but then I get a 0,0 location with a good refresh rate...quite useless. This seems to be coming from Windows, like in the previous forum post, W seems to remove the driver from ublox and replaces it with it's own to read the data which pooches a good reading on old tablet that were upgraded...

In u-blox center, there is a place to update the firmware, but from my readings we might not be able to actually update it or downgrade it to make it work.

**Comment by ghuotvez** [comment]
Alright I think I have something that works, not the easiest way of doing though... 

Basically the app can't suddenly make Windows work right with the ublox receiver in our old Panasonic tablets, but it should work fine without the following on brand new tablets. We need some tweaking in order to achieve this for older ones. Here's the recipe to test on those other tablets.

1. Uninstall u-blox sensor, this will force the tablet to use the latest driver from Windows.
*  Follow the guide in here: https://www.u-blox.com/sites/default/files/WindowsDrivers_AppNote_%28UBX-19005821%29.pdf
*  This will make you uninstall the associated port of u-blox and the sensor itself. After reboot in Device Manager > Ports, we should see USB Serial Device.
2. Force the tablet to read the u-blox receiver from a given com port
*  Download GPSComplete x64: https://www.turboirc.com/update2/tu.php?p=99fc8ed0-fafa-4bd9-9f10-0c923392dd59&f=99fc8ed0-fafa-4bd9-9f10-0c923392dd51
*  In drivers tab, select "Install Driver" for GPSDirect Driver
*  NMEA source --> COM3 (or the port # you see in Device Manager > Port > USB Serial Device)
*  Check all boxes below
3.  Install latest release of GSC Field App, there is a small code change in the call for location that helps to make sure wifi connection is out of the way since it's the default for Windows and if location doesn't budge that much it'll revert straight to it...

I'll test this on another tablet for good measure, but I was able to make it work and get my location with a 1m accuracy without any connected bluetooth gps or wifi or anything else.  I was also able to make the app work good even with GPSComplete closed, as long as the driver has been installed for COM3 port this should work. 

_Warning: this will make Panasonic Viewer stop working. I recommend downloading ublox center instead: https://www.u-blox.com/sites/default/files/u-centersetup_v21.02.exe.zip_

**Comment by jameljoseph** [comment]
Sweet I will test this out!!!

**Comment by ghuotvez** [comment]
Did some testing with my latest built and the magic recipe above. I was skeptical at first because the accuracy was always 1 meter. Couldn't get something around 3-5m like it used to be on this tablet. Went out for a quick drive on an highway, a tight bridge full of metal columns and wires and the points are right on spot. I can even see on which side of the highway I was driving and same thing which side of the bridge the points were taken, even though it's a tight one with only 1 lane per side.

I noticed some weird behavior though within the map page. Lost connection of the GPS at some point then the point went to 0.0 and never got back and froze the whole thing.

**Comment by ghuotvez** [comment]
Like I thought, when accuracy goes bad it is set to 1 rather then 0 ... not sure why but it might be coming from the driver installed with gpscomplete not necessarily coming from the tablet receiver.

**Comment by ghuotvez** [comment]
From @saralizj 

> I’m happy to report that our GPS coordinates have low error and see to be spot on.  This was a big concern and I’m thrilled to see we have no issues there.

That's coming from brand view panasonic tablets. So that's a great news.

**Comment by ghuotvez** [comment]
I went back to the office and took another tablet for testing. This one was in W10 1609 build. I've updated it up to 20H2 then I installed my latest testing app and it worked fine. The very old in 1609 was also working before upgrading both the OS and the app. And this time, since I didn't install the custom driver from gpscomplete I don't have an accuracy problem jammed on 1m. I get a changing one that goes from 6 to 7m.

So final comment, it'll depend on the tablet and it's behavior to know which one is a preferable method, new app only or new app with gpscomplete and uninstall of old receiver drivers.
